Make DispatchKeyExtractor forget about TensorOptions

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#36290 Make DispatchKeyExtractor forget about TensorOptions**

The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[Codegen diff](https://github.com/bhosmer/scratch/compare/before_36290...after_36290)

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

Commit:
Make DispatchKeyExtractor forget about TensorOptions

The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[ghstack-poisoned]

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[ghstack-poisoned]

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

[ghstack-poisoned]

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[Codegen diff](https://github.com/bhosmer/scratch/compare/before_36290...after_36290)

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

[ghstack-poisoned]

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[Codegen diff](https://github.com/bhosmer/scratch/compare/before_36290...after_36290)

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

[ghstack-poisoned]

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[Codegen diff](https://github.com/bhosmer/scratch/compare/before_36290...after_36290)

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

[ghstack-poisoned]

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[Codegen diff](https://github.com/bhosmer/scratch/compare/before_36290...after_36290)

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

[ghstack-poisoned]

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[Codegen diff](https://github.com/bhosmer/scratch/compare/before_36290...after_36290)

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

[ghstack-poisoned]

smessmer commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -509,7 +512,7 @@ def __getitem__(self, x):
     'formals_types': List[str],
     'formals_types_with_return': List[str],
-    'inferred_key_set': str,
+    # 'inferred_key_set': str,
```
why not remove entirely?

bhosmer replied:
doh

smessmer reviewed (approved):
You can also remove the TODO in DispatchKeyExtractor::getDispatchKeyBoxed

Commit:
Update on "Make DispatchKeyExtractor forget about TensorOptions"


The BackendSelect dispatch key gives us a way to extract backend-
specific dispatch keys from non-Tensor arguments without teaching
the DispatchKeyExtractor about them. Here we finish switching over
to the BackendSelect approach for factory functions and remove
TensorOptions from the set of types DispatchKeyExtractor needs to
consider.

[Codegen diff](https://github.com/bhosmer/scratch/compare/before_36290...after_36290)

Differential Revision: [D20936595](https://our.internmc.facebook.com/intern/diff/D20936595)

[ghstack-poisoned]

kostmo commented:
Looks like this PR broke `pytorch_bazel_test` in master.  Unfortunately, the failure was masked due to the fact that `pytorch_bazel_test` job was broken in master a few days ago (at the merge base of this PR).

Reverting now.
